### PR TITLE
Improvements of integer overflow checks

### PR DESCRIPTION
**Pull request by zilder**
1. The `INTPL()` has a problem. When second and third arguments are the same variable, overflow check doesn't work. E.g.:
```
INTPL(pairs[index].val, clamp_sum, clamp_sum);
```
This PR fixes this and also introduces an optimization for postgres 11+ and gcc 5+ by using `__builtin_add_overflow`.

2. Second improvement relates to the parser. This PR replaces complex set of conditions that check for integer overflow with a single one (checking `errno` after calling `strtol()`).

### CODE REVIEW

File: src/is_parser.c [hunk 1]
```diff
@@ -82,6 +75,10 @@ is_parse(ISParser *parser)
             SKIP_SPACES(parser->ptr);
             SKIP_ESCAPED(parser->ptr);
             GET_NUM(parser->ptr, key, parser->begin);
+
+            if (key < PG_INT32_MIN || key > PG_INT32_MAX)
```
Comment (za-arthur) [inline src/is_parser.c:79]: `GET_NUM` is used below for another `key` variable in `is_parse_arr()` function which is still `int32`. Probably it is better to change its type as well and check for overflow or just add new macros for `int32` types.

File: src/istore.h [hunk 2]
```diff
@@ -150,12 +150,21 @@ int is_int32_arr_comp(const void *a, const void *b);
...
-    do                                                                                                                 \
-    {                                                                                                                  \
-        _r = _a + _b;                                                                                                  \
-        if (SAMESIGN(_a, _b) && !SAMESIGN(_r, _a))                                                                     \
+    do {                                                                                                               \
+        if (__builtin_add_overflow(_a, _b, &_r))                                                                       \
             ereport(ERROR, (errcode(ERRCODE_NUMERIC_VALUE_OUT_OF_RANGE), errmsg("integer out of range")));             \
     } while (0)
+#else
+#define INTPL(_a, _b, _r)                                                                                              \
```
Comment (za-arthur) [inline src/istore.h:161]: I think you can reuse `pg_add_s32_overflow()` and `pg_add_s64_overflow()` and delete `INTPL()` macros for good.

```
if (pg_add_s32_overflow(pairs[index].val, clamp_sum, &clamp_sum))
    ereport(ERROR, (errcode(ERRCODE_NUMERIC_VALUE_OUT_OF_RANGE), errmsg("integer out of range")));
```

**Review by za-arthur — COMMENTED** [review]
Generally looks good! There are just couple comments.

File: src/is_parser.c (hunk 1 above)
Comment (zilder) [reply src/is_parser.c:79]: True, i didn't notice the `is_parse_arr()` func. I'll fix it.

File: src/istore.h (hunk 2 above)
Comment (zilder) [reply src/istore.h:161]: That was my intention, but then CI reported failures. Turned out those functions were introduced in pg11. We gotta keep supporting older version (at least up to pg10).

File: src/is_parser.c (hunk 1 above)
Comment (zilder) [reply src/is_parser.c:79]: I fixed this. Also I regenerated test scripts (using dumbo) which caused a lot of noise (changed line numbers).

**Review by za-arthur — APPROVED** [review]
Thanks. Looks good to me.

**Pull request merged**